jbpm-flow/src/main/java/org/jbpm/process/core/context/variable/VariableScope.java

```diff
@@ -102,14 +102,17 @@ public Object validateVariable(String processName, String name, Object value) {
             }
         } else if (value != null) {
             DataType type = var.getType();
+            ObjectDataType obj = new ObjectDataType();
             boolean isRightType = type == null || type.verifyDataType(value);
             if (!isRightType ) {
                 if (variableStrictEnabled) {
                     throw new IllegalArgumentException("Variable '" + name + "' has incorrect data type expected:" + var
                             .getType().getStringType() + " actual:" + value.getClass().getName());
-                } else if (value instanceof String) {
+                } else if (value instanceof String && (type.getClass() != obj.getClass())) {
```

Comment:
Probably you can write it more clear like this
```suggestion
                } else if (value instanceof String) {
                    if (type.getClass() == obj.getClass()) {
                        value = type.readValue(value.toString());
                    else {
                        value = type.valueOf(value.toString());
                    }
                }
```